File: mirea/cources/2026-2027/aii-august/03/contest/C.py
inf = 1 << 60
def solve():
    # читаем данные:
    n, k = map(int, input().split())
    a = list(map(int, input().split()))
    # сортируем участников по возрастанию их результатов:
    a.sort()
    # делим участников на группы
    dp = [[inf] * (k + 1) for _ in range(n + 1)]
    dp[0][0] = 0
    for i in range(n):
        for sz in range(2, 5):
            j = i - sz + 1
            if j < 0:
                continue
            # группа [j, i] стоимостью a[i] - a[j]
            for wasGroups in range(k):
                dp[i + 1][wasGroups + 1] = min(
                    dp[i + 1][wasGroups + 1],
                    dp[j][wasGroups] + a[i] - a[j]
                )
    # выводим ответ:
    print(dp[n][k])

tt = int(input())
while tt > 0:
    solve()
    tt -= 1

File: mirea/cources/2026-2027/aii-august/03/contest/C.cpp
#include <bits/stdc++.h>
#include <cassert>
#define all(x) (x).begin(),(x).end()
using namespace std;
using ll = long long;
const ll inf = 1LL << 60;
void setmin(auto &x, const auto &y) { if (x > y) x = y; }
void solve() {
    // читаем данные:
    int n, k; cin >> n >> k;
    vector<int> a(n);
    for (auto &it : a)
        cin >> it;
    // сортируем участников по возрастанию их результатов:
    sort(all(a));
    // делим участников на группы
    vector dp(1+n, vector(k+1, +inf));
    dp[0][0] = 0;
    for (int i = 0; i < n; i++) {
        for (int sz = 2; sz <= 4; sz++) {
            int j = i - sz + 1;
            if (j < 0) continue;
            // группа [j, i] стоимостью a[i] - a[j]
            for (int wasGroups = 0; wasGroups < k; wasGroups++)                
                setmin(dp[i+1][wasGroups+1], dp[j][wasGroups] + a[i] - a[j]);
        }
    }
    // выводим ответ:
    cout << dp[n][k] << '\n';
}
main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    int tt; cin >> tt;
    while (tt --> 0) solve();
}
